pkg/controllers/cloud/vpc_cache.go

```diff
@@ -0,0 +1,122 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloud
+
+import (
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+
+	runtimev1alpha1 "antrea.io/nephe/apis/runtime/v1alpha1"
+)
+
+type VpcCacheData struct {
+	VpcIndexer cache.Indexer
+}
+
+var VpcCache *VpcCacheData
+
+const (
+	VpcIndexerByAccountNameSpacedName = "namespace.accountname"
+	VpcIndexerByAccountUUID           = "vpc.uuid"
+	VpcIndexerByNamespace             = "vpc.namespace"
+	VpcIndexerByName                  = "vpc.Name"
+	VpcIndexerByVpcID                 = "vpc.id"
+)
+
+func init() {
+	VpcCache = &VpcCacheData{}
+}
+
+// DefineIndexers defines an indexer.
+func DefineIndexers() {
+	VpcCache.VpcIndexer = cache.NewIndexer(
+		func(obj interface{}) (string, error) {
+			vpc := obj.(*runtimev1alpha1.Vpc)
+			return fmt.Sprintf("%v-%v", vpc.Annotations[VpcIndexerByAccountUUID], vpc.Info.Id), nil
+		},
+		cache.Indexers{
+			VpcIndexerByAccountNameSpacedName: func(obj interface{}) ([]string, error) {
+				vpc := obj.(*runtimev1alpha1.Vpc)
+				return []string{vpc.Annotations[VpcIndexerByAccountNameSpacedName]}, nil
+			},
+			VpcIndexerByVpcID: func(obj interface{}) ([]string, error) {
+				vpc := obj.(*runtimev1alpha1.Vpc)
+				return []string{vpc.Info.Id}, nil
+			},
+			VpcIndexerByNamespace: func(obj interface{}) ([]string, error) {
+				vpc := obj.(*runtimev1alpha1.Vpc)
+				return []string{vpc.Namespace}, nil
+			},
+			VpcIndexerByName: func(obj interface{}) ([]string, error) {
+				vpc := obj.(*runtimev1alpha1.Vpc)
+				return []string{vpc.Name}, nil
+			},
+		})
+}
+
+// BuildVpcInventory using vpc list from cloud updates global vpcCache(with vpcs applicable for the current account).
+func BuildVpcInventory(vpcList map[string]*runtimev1alpha1.Vpc, namespacedName *types.NamespacedName,
+	uuid string, log logr.Logger) error {
+	vpcsInCache, err := VpcCache.VpcIndexer.ByIndex(VpcIndexerByAccountNameSpacedName, namespacedName.String())
+	if err != nil {
+		return fmt.Errorf("fetching vpc list from global cache failed, "+
+			"index %v, error %v", namespacedName.String(), err)
+	}
+
+	// Remove vpcs in internal global cache which are not found in vpc list fetched from cloud.
+	for _, i := range vpcsInCache {
+		vpc := i.(*runtimev1alpha1.Vpc)
+		if _, found := vpcList[vpc.Info.Id]; !found {
+			log.Info("Deleting a vpc from global cache", "vpc id", vpc.Info.Id, "account",
+				vpc.Annotations[VpcIndexerByAccountNameSpacedName])
+			if err := VpcCache.VpcIndexer.Delete(vpc); err != nil {
+				log.Error(err, "delete from global vpc cache failed", "indexer", namespacedName.String())
```

Comment:
In the log message, indexer key does not make sense here, as you are dumping namespaceName